### ISSUE TITLE
Sobel Edge Detection

### ISSUE THREAD

**Issue by BensonBen** [lovell/sharp#3147]
### Applying Sobel Edge Detection to Images applying machine learning techniques.

I've been working on a personal project that supports automatic coloring book page generation. 
Started off using sharp for some of its major features. Conversion, optimization, cropping, blur, grayscale etc.
Then when it came time to use something like Sobel or Canny on the images I didn't have too many options.
Eventually, I made my own implementation which I've released into the wild. 

- [github link](https://github.com/BensonBen/musical-sniffle)
- [published npm library](https://www.npmjs.com/package/@musical-sniffle/sobel-edge-detection)

### There wasn't any features related to edge detection that I could find.
n/a

### What would you expect the API to look like?

This change would involve wrapping, grayscale, blur and the suggested Sobel's algorithm to an image.
Current implementations look like.
```typescript
// using this implementation requires the ability to create a 
import * as sharp from 'sharp';
import { SobelService } from '@musical-sniffle/sobel-edge-detection';

const sobelService = new SobelService();
// sobel's algorithm uses grayscale image data.
const { data, info } = await sharp(`${__dirname}/my-image.png`)
  .ensureAlpha()
  .grayscale()
  // optional, but useful to do blur beforehand.
  // .blur(3)
  .raw()
  .toBuffer({ resolveWithObject: true });

const { width, channels, height } = info;
const { imageData: detected } = sobelService.applySobel(
  new Uint8ClampedArray(data.buffer),
  width,
  height,
  channels
);
const imagename = `${__dirname}/edge-detected-image.png`;
await sharp(detected, { raw: { channels: 4, height, width } }).toFile(
  imagename
);
```

Future Implementations may look like
```typescript
// using this implementation requires the ability to create a 
import * as sharp from 'sharp';

const { data, info } = await sharp(`${__dirname}/my-image.png`)
  .applySobel({ blur: 3, kernelSize: 3 })
  .toBuffer({ resolveWithObject: true });

const imagename = `${__dirname}/edge-detected-image.png`;
await sharp(detected, { raw: { channels: 4, height, width } }).toFile(
  imagename
);
```

### I've implemented it myself and have successfully used it with sharp.

see above links to github and my public npm repository. Since my implementation uses some relatively low level constructs
to apply sobel to an image. We could leverage what I've already written to make this process a bit more seemless.

### Images for Context.
| Image         |
|--------------|
| ![audi-r8-v10-2020-rwd-c258820032022223329_1](https://user-images.githubusercontent.com/10961012/159187239-8ea343c0-5d8b-4f00-9de5-7fed7a6c8fa6.jpg) |
| ![download](https://user-images.githubusercontent.com/10961012/159187246-45feed3d-e732-4961-bb2e-6af2859ef886.png) |

**Comment by lovell** [comment]
Hi, did you see the [convolve](https://sharp.pixelplumbing.com/api-operation#convolve) operation? The docs contain an example of using it with a Sobel kernel.

**Comment by BensonBen** [comment]
nope, I didn't see that and now I feel pretty dumb. Is there a way to do canny edge detection then? While convolve works for applying a sobel operation to an image. It would seem like you'd also need Theta to produce the "next step" for canny. I didn't see a way to chain operators together to produce that. Speaking of non-max suppression and hysteresis.

I think non-max suppression could be done if I could get Theta back from the convolve (or another) operator.

Thanks for your response!

**Comment by lovell** [comment]
libvips provides both Canny and Sobel as dedicated operations but these are not yet exposed in sharp. Happy to accept a PR, if you're able.

**Comment by BensonBen** [comment]
@lovell definitely going to explore this option. Haven't done any NodeJS to C++ bindings before, but I'll do my best. TLDR: trying to work towards a PR

**Comment by lovell** [comment]
Thanks, perhaps look at the existing `convolve` logic to understand how it passes values from JS to C++. I'm always happy to review WIP code and/or answer questions.

**Comment by shahkashani** [comment]
For what it's worth, I am cheering for this to happen!

**Comment by baparham** [comment]
I might be interested in this. @BensonBen are you still planning on implementing a PR? It's not too terrible to expose a libvips function (at least the one time I tried to in #2726 which maybe would help show one way of doing it) in the c++ node bindings. Libvips has pretty good documentation about their c api that you are exposing, and then you basically just sort of copy function definitions up the stack and insert a call to it in the image processing pipeline (`./src/pipeline.cc`).

**Comment by BensonBen** [comment]
@baparham I tried doing this when I had more freetime since it's been a while (undergrad 5 years ago) since I've done any c++ development. That being said I was able to get canny edge detection working. However, when it came to figuring out how the operators are applied `pipeline.cc` i.e. precedence and javascript method mappings. it seemed like a mess and I gave up

**Comment by lukejans** [comment]
I just started using sharp yesterday and I've been looking at code longer then my eyes can handle trying to figure out how to retrieve values from a Sobel operation using convolve that are outside of the range 0-255. This issue and issue #3455 seemed to be the most similar so I decided to drop a comment here to try to get some help with this. 

This is the code I'm running and when reading the pixel data returned it all seems to stay within the bounds of 0-255 but I need the values to be true to the original product of the Sobel operation so I can calculate the angle of the vectors.
```js
import fs from "node:fs/promises";
import sharp from "sharp";

const imgBuffer = await fs.readFile("testing/test.png");

const gx = await sharp(imgBuffer)
    .resize(imgSize)
    .greyscale()
    .blur(1.0)
    .convolve({
        width: 3,
        height: 3,
        kernel: [-1, 0, 1, -2, 0, 2, -1, 0, 1],
    })
    .raw({ depth: "short" })
    .toBuffer({ resolveWithObject: true });

console.log(gx.data);
```
Not sure if I'm missing something but hopefully there is a solution to this. Nonetheless, this is an incredible tool and I'm very grateful for it's existence :)

**Comment by lovell** [comment]
@lukejans If your input is 16-bit PNG then try [pipelineColourspace](https://sharp.pixelplumbing.com/api-colour/#pipelinecolourspace) with `rgb16` to retain 16-bit throughout. If you require further help, please open a new question.
